lua/ttt2/libraries/orm.lua

```diff
@@ -0,0 +1,219 @@
+---
+-- Object Relational Model
+-- @author Histalek
+
+if SERVER then
+	AddCSLuaFile()
+end
+
+local sql = sql
+
+orm = orm or {}
+
+local ORMMODEL = {}
+
+---
+-- Returns an object relational model according to the specified databasetable. Does nothing if no databasetable with the given name exists.
+-- @param string tableName The name of the table in the database to create a model for.
+-- @param[opt] boolean force If set to `true` the function will not return a cached version of the model.
+-- @return ormmodel The model of the database table.
+-- @realm shared
+function orm.Make(tableName, force)
+	if IsValid(orm[tableName]) and not force then
+		return orm[tableName]
+	end
+
+	if not sql.TableExists(tableName) then return end
+
+	local model = {}
+	local primaryKey = sql.GetPrimaryKey(tableName)
+	local dataStructure = sql.GetTableColumns(tableName)
+
+	model._tableName = tableName
+	model._primaryKey = primaryKey
+	model._dataStructure = dataStructure
+	model.All = ORMMODEL.All
+	model.Find = ORMMODEL.Find
+	model.New = ORMMODEL.New
+
+	-- DO NOT setup delete/save functions if no primarykey is found.
+	-- In those cases the 'rowid' column would function as the primarykey, but as the rowid could change anytime (https://www.sqlite.org/rowidtable.html) data could be deleted unintentionally.
+	-- Most likely rowids wont change in gmod as there is no vacuum operation but just to be safe we will not allow to use such tables. ref: https://wiki.facepunch.com/gmod/sql
+
+	if not primaryKey then
```

Comment:
Does this work, if sql.GetPrimaryKey() always returns a table?

Reply:
from the docs of `sql.GetPrimaryKey()`
```lua
---
-- Returns the primarykey column names of the specified table in order of their index.
-- @param string tableName The name of the table to search.
-- @return table|nil Returns a table of the primarykey columns.
-- @realm shared
```
it does return nil in case of no primarykey being set